packages/ui-components/src/components/flow-canvas/canvas-context.tsx

```diff
@@ -103,17 +105,111 @@ export const InteractiveContextProvider = ({
   const selectedFlowActionRef = useRef<Action | null>(null);
   const selectedNodeCounterRef = useRef<number>(0);
   const state = useStoreApi().getState();
+  const [actionToPaste, setActionToPaste] = useState<Action | null>(null);
 
   const spacePressed = useKeyPress(SPACE_KEY);
   const shiftPressed = useKeyPress(SHIFT_KEY);
 
-  const canvas = useMemo(() => {
-    return flowCanvasContainerId
-      ? document.getElementById(flowCanvasContainerId)
-      : null;
-  }, [flowCanvasContainerId]);
-  const copyPressed = useKeyPress(COPY_KEYS, { target: canvas });
-  const { fetchClipboardOperations } = useClipboardContext();
+  const handleCopy = useCallback((action: Action, actionCount: number) => {
+    const flowString = JSON.stringify(action);
+
+    navigator.clipboard
+      .writeText(flowString)
+      .then(() => {
+        setActionToPaste(action);
+        copyPasteToast({
+          success: true,
+          isCopy: true,
+          itemsCount: actionCount,
+        });
+      })
+      .catch(() => {
+        copyPasteToast({
+          success: false,
+          isCopy: true,
+          itemsCount: actionCount,
+        });
+      });
+  }, []);
+
+  const copySelectedStep = useCallback(() => {
+    if (!selectedStep) {
+      return;
+    }
+
+    const stepDetails = flowHelper.getStep(flowVersion, selectedStep);
+
+    if (!stepDetails || !flowHelper.isAction(stepDetails.type)) {
+      return;
+    }
+
+    const stepToBeCopied = cloneDeep(stepDetails);
+    stepToBeCopied.nextAction = undefined;
+    flowHelper.clearStepTestData(stepToBeCopied);
+
+    handleCopy(stepToBeCopied as Action, 1);
+  }, [flowVersion, handleCopy, selectedStep]);
+
+  const copySelectedArea = useCallback(() => {
+    const selectionArea = document.querySelector(
+      `.${NODE_SELECTION_RECT_CLASS_NAME}`,
+    );
+
+    if (!selectionArea) {
+      selectedFlowActionRef.current = null;
+      selectedNodeCounterRef.current = 0;
+      return;
+    }
+    if (!selectedFlowActionRef.current || !selectedNodeCounterRef.current) {
+      return;
+    }
+
+    handleCopy(selectedFlowActionRef.current, selectedNodeCounterRef.current);
+  }, [handleCopy]);
+
+  useEffect(() => {
+    if (!flowCanvasContainerId) return;
+
+    const copyHandler = (event: ClipboardEvent) => {
+      event.preventDefault();
+      if (selectedStep) {
+        copySelectedStep();
+      } else {
+        copySelectedArea();
+      }
+    };
+
+    const isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
+    let targetEl: HTMLElement | Document | null = null;
+
+    if (isSafari) {
+      document.addEventListener('copy', copyHandler);
```

Comment:
Safari copy handler does not work when attaching to the canavs element. Here we should probably check in the event itseft and not prevent default if copy is from outside canvas